core/modules/simpletest/simpletest.module

```diff
@@ -364,28 +364,52 @@ function simpletest_test_get_all() {
     else {
       $files = backdrop_system_listing('/^' . BACKDROP_PHP_FUNCTION_PATTERN . '\.tests\.info$/', 'modules', 'name', 0);
       $groups = array();
+      $check_keys = array('name', 'description', 'group', 'file');
       foreach ($files as $file) {
         $classes = backdrop_parse_info_file($file->uri, TRUE);
-        // If the .tests.info file contained no classes, then it puts a slew of
-        // hard-to-diagnose errors into the watchdog log when the testing page
-        // is loaded. So we'll check and put a more informative error here.
-        $class_values = array_values($classes);
-        if (empty($classes) || !is_array($class_values[0])) {
-          watchdog('simpletest', 'File %uri contains no testing classes.', array('%uri' => $file->uri), WATCHDOG_WARNING, l(t('See the Testing Framework documentation.'), 'https://docs.backdropcms.org/documentation/testing-framework'));
-          continue;
+        $error = '';
+        if (empty($classes)) {
+          $error = 'File %uri contains no testing classes.';
         }
-        foreach ($classes as $class => $info) {
-          $info['file path'] = dirname($file->uri);
-          // If this test class requires a non-existing module, skip it.
-          if (!empty($info['dependencies'])) {
-            foreach ($info['dependencies'] as $module) {
-              if (!backdrop_get_filename('module', $module)) {
-                continue 2;
+        else {
+          foreach ($classes as $class => $info) {
+            // If the .tests.info file is malformed, it puts a slew of
+            // hard-to-diagnose errors into the watchdog log when the testing page
+            // is loaded. So we'll check for various malformalities and put a more
+            // informative error.
+            if (!is_array($info) || empty($info)) {
+              $error = format_string('File %uri contains a definition for %class that is not an array of testing class fields.', array('%class' => $class));
+              continue;
+            }
+            else {
+              foreach ($check_keys as $key) {
+                if (!isset($info[$key])) {
+                  $error = format_string('File %uri contains a definition for %class that is missing field %key.', array('%class' => $class, '%key' => $key));
```

Comment:
The logic seems to be bailing out soon as a single error is found for any of the `$check_keys`. I'd expect an `$keys` (plural) array, so that if a .tests.info file has multiple malformities, then we warn about all of them. So I'd also expect a `format_plural()` here, checking on the count of `$keys`, and if multiple, then say "...is missing the [implode()'d $keys] fields." for the plural variant.

Reply:
Definitely the former. Putting multiple watchdog messages per file for each detected malformation would be pretty annoying. In a similar vein, if there are multiple malformed classes in the file, we don't put a message for each malformation detected (especially because once `backdrop_parse_info()` gets confused, all bets are off on what might show up in the parsed array). But reporting if multiple fields are missing is a nice courtesy, so added.